File: public/api/panel/convenios.php
<?php
ini_set('display_errors', 0);
error_reporting(0);

// Carrega o autoloader do Composer
require_once __DIR__ . '/../../../vendor/autoload.php';

// Carrega as variáveis de ambiente
$envPath = __DIR__ . '/../../../.env';
if (file_exists($envPath)) {
    $dotenv = Dotenv\Dotenv::createImmutable(__DIR__ . '/../../../');
    $dotenv->load();
}

// Usa a classe Database para conexão
use App\Config\Database;

try {
    $pdo = Database::getConnection();
} catch (PDOException $e) {
    http_response_code(500);
    echo json_encode(['error' => 'Erro na conexão com banco de dados']);
    exit;
}


header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    exit(0);
}

try {
    
    if ($_SERVER['REQUEST_METHOD'] === 'GET') {
        // Buscar todos os convênios
        $stmt = $pdo->prepare("SELECT * FROM convenios_aceitos ORDER BY nome");
        $stmt->execute();
        $convenios = $stmt->fetchAll(PDO::FETCH_ASSOC);
        
        echo json_encode([
            'success' => true,
            'convenios' => $convenios
        ]);
        
    } elseif ($_SERVER['REQUEST_METHOD'] === 'POST') {
        // Criar novo convênio
        $input = json_decode(file_get_contents('php://input'), true);
        
        if (!$input || !isset($input['nome'])) {
            throw new Exception('Nome do convênio é obrigatório');
        }
        
        $stmt = $pdo->prepare("
            INSERT INTO convenios_aceitos (
                nome, registro_ans, observacoes, ativo
            ) VALUES (?, ?, ?, ?)
        ");
        
        $stmt->execute([
            $input['nome'],
            $input['registro_ans'] ?? null,
            $input['observacoes'] ?? null,
            1 // Sempre ativo ao criar
        ]);
        
        echo json_encode([
            'success' => true,
            'message' => 'Convênio criado com sucesso',
            'id' => $pdo->lastInsertId()
        ]);
        
    } elseif ($_SERVER['REQUEST_METHOD'] === 'PUT') {
        // Atualizar convênio (suporte a atualizações parciais)
        $input = json_decode(file_get_contents('php://input'), true);
        
        if (!$input || !isset($input['id'])) {
            throw new Exception('ID é obrigatório');
        }
        
        // Construir query dinamicamente baseada nos campos fornecidos
        $updateFields = [];
        $params = [];
        
        if (isset($input['nome'])) {
            $updateFields[] = 'nome = ?';
            $params[] = $input['nome'];
        }
        
        if (isset($input['registro_ans'])) {
            $updateFields[] = 'registro_ans = ?';
            $params[] = $input['registro_ans'];
        }
        
        if (isset($input['observacoes'])) {
            $updateFields[] = 'observacoes = ?';
            $params[] = $input['observacoes'];
        }
        
        if (isset($input['ativo'])) {
            $updateFields[] = 'ativo = ?';
            $params[] = $input['ativo'];
        }
        
        if (empty($updateFields)) {
            throw new Exception('Nenhum campo para atualizar foi fornecido');
        }
        
        // Adicionar updated_at e ID para WHERE
        $updateFields[] = 'updated_at = NOW()';
        $params[] = $input['id']; // ID para WHERE
        
        $sql = "UPDATE convenios_aceitos SET " . implode(', ', $updateFields) . " WHERE id = ?";
        $stmt = $pdo->prepare($sql);
        $stmt->execute($params);
        
        echo json_encode([
            'success' => true,
            'message' => 'Convênio atualizado com sucesso'
        ]);
        
    } elseif ($_SERVER['REQUEST_METHOD'] === 'DELETE') {
        // Deletar convênio
        $input = json_decode(file_get_contents('php://input'), true);
        
        if (!$input || !isset($input['id'])) {
            throw new Exception('ID do convênio é obrigatório');
        }
        
        $stmt = $pdo->prepare("DELETE FROM convenios_aceitos WHERE id = ?");
        $stmt->execute([$input['id']]);
        
        if ($stmt->rowCount() === 0) {
            throw new Exception('Convênio não encontrado');
        }
        
        echo json_encode([
            'success' => true,
            'message' => 'Convênio deletado com sucesso'
        ]);
        
    } else {
        throw new Exception('Método não permitido');
    }
    
} catch (Exception $e) {
    echo json_encode([
        'success' => false,
        'message' => $e->getMessage()
    ]);
} 

File: app/routers/panel/convenios.py
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import JSONResponse
from fastapi.encoders import jsonable_encoder
from ...core.db import get_db, is_postgres_connection
from ..auth import verify_admin_user
from ...schemas.panel import ConvenioCreate, ConvenioUpdate

router = APIRouter(prefix="/panel", tags=["panel-convenios"], dependencies=[Depends(verify_admin_user)])

@router.get("/convenios")
async def listar_convenios():
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            with db.cursor() as cur:
                cur.execute("SELECT * FROM convenios_aceitos ORDER BY nome")
                convenios = cur.fetchall()
                return JSONResponse(content=jsonable_encoder({"success": True, "convenios": convenios}))
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": "Erro ao buscar convênios", "error": str(e)}, status_code=500)


@router.post("/convenios")
async def criar_convenio(payload: ConvenioCreate):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            with db.cursor() as cur:
                if is_postgres_connection(db):
                    cur.execute(
                        "INSERT INTO convenios_aceitos (nome, registro_ans, observacoes, ativo) VALUES (%s, %s, %s, %s) RETURNING id",
                        (payload.nome, payload.registro_ans, payload.observacoes, 1 if payload.ativo else 0)
                    )
                    result = cur.fetchone()
                    convenio_id = result["id"] if isinstance(result, dict) else result[0]
                else:
                    cur.execute(
                        "INSERT INTO convenios_aceitos (nome, registro_ans, observacoes, ativo) VALUES (%s, %s, %s, %s)",
                        (payload.nome, payload.registro_ans, payload.observacoes, 1 if payload.ativo else 0)
                    )
                    convenio_id = cur.lastrowid
                
                return JSONResponse(content={"success": True, "message": "Convênio criado com sucesso", "id": convenio_id})
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": f"Erro ao criar convênio: {str(e)}"}, status_code=500)


@router.put("/convenios/{convenio_id}")
async def atualizar_convenio(convenio_id: int, payload: ConvenioUpdate):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            updates = []
            values = []
            
            for field, value in payload.dict(exclude_unset=True).items():
                if value is not None:
                    updates.append(f"{field} = %s")
                    if field == "ativo":
                        values.append(1 if value else 0)
                    else:
                        values.append(value)
            
            if not updates:
                return JSONResponse(content={"success": False, "message": "Nenhum campo para atualizar"}, status_code=400)
            
            updates.append("updated_at = CURRENT_TIMESTAMP")
            values.append(convenio_id)
            
            with db.cursor() as cur:
                cur.execute("SELECT id FROM convenios_aceitos WHERE id = %s", (convenio_id,))
                if not cur.fetchone():
                    return JSONResponse(content={"success": False, "message": "Convênio não encontrado"}, status_code=404)
                
                query = f"UPDATE convenios_aceitos SET {', '.join(updates)} WHERE id = %s"
                cur.execute(query, values)
                
                return JSONResponse(content={"success": True, "message": "Convênio atualizado com sucesso"})
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": f"Erro ao atualizar convênio: {str(e)}"}, status_code=500)


@router.delete("/convenios/{convenio_id}")
async def deletar_convenio(convenio_id: int):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            with db.cursor() as cur:
                cur.execute("SELECT id FROM convenios_aceitos WHERE id = %s", (convenio_id,))
                if not cur.fetchone():
                    return JSONResponse(content={"success": False, "message": "Convênio não encontrado"}, status_code=404)
                
                cur.execute("UPDATE convenios_aceitos SET ativo = 0 WHERE id = %s", (convenio_id,))
                return JSONResponse(content={"success": True, "message": "Convênio desativado com sucesso"})
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": f"Erro ao deletar convênio: {str(e)}"}, status_code=500)


@router.get("/convenios/{convenio_id}")
async def obter_convenio(convenio_id: int):
    try:
        db_gen = get_db()
        db = next(db_gen)
        try:
            with db.cursor() as cur:
                cur.execute("SELECT * FROM convenios_aceitos WHERE id = %s", (convenio_id,))
                convenio = cur.fetchone()
                if not convenio:
                    return JSONResponse(content={"success": False, "message": "Convênio não encontrado"}, status_code=404)
                
                return JSONResponse(content=jsonable_encoder({"success": True, "convenio": convenio}))
        finally:
            try:
                db.close()
            except Exception:
                pass
    except Exception as e:
        return JSONResponse(content={"success": False, "message": f"Erro ao buscar convênio: {str(e)}"}, status_code=500)


# Endpoints para compatibilidade
@router.put("/convenios")
async def atualizar_convenio_legacy(payload: dict):
    if not payload or "id" not in payload:
        raise HTTPException(status_code=400, detail="ID é obrigatório")
    
    convenio_id = payload.pop("id")
    convenio_update = ConvenioUpdate(**payload)
    return await atualizar_convenio(convenio_id, convenio_update)


@router.delete("/convenios")
async def deletar_convenio_legacy(payload: dict):
    if not payload or "id" not in payload:
        raise HTTPException(status_code=400, detail="ID do convênio é obrigatório")
    
    return await deletar_convenio(payload["id"])